### PR TITLE
Improved readme

### PR DESCRIPTION
**Pull request by alexios-angel**
I added compiler explorer links as well as some better examples.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,10 +4,27 @@
...
+   - [C++17 syntax](#C++17-syntax)
+   - [C++20 syntax](#C++20-syntax)
+ - [Examples](#Examples)
+   - [Extracting number from input](#Extracting-number-from-input)
+   - [Extracting-values-from-date](#Extracting-values-from-date)
+   - [Using captures](#Using-captures)
+   - [Lexer](#Lexer)
+   - [Range over input](#Range-over-input)
+   - [Unicode](#Unicode)
+ - [Integration](#Integration)
```
Comment (friendlyanon) [inline README.md:27]: ToC is already provided by GitHub's UI.

File: README.md [hunk 2]
```diff
@@ -247,35 +295,80 @@ std::optional<lex_item> lexer(std::string_view v) noexcept {
...
+#include "ctre.hpp"
+
+//using namespace ctre;
+```
+This header can be regenerated with `make single-header`.
+### CMake
+If you are using cmake, you can add this directory as subdirectory and link to target `ctre`.
+
+```CMake
+cmake_minimum_required(VERSION 3.8.0)
```
Comment (friendlyanon) [inline README.md:352]:
```suggestion
cmake_minimum_required(VERSION 3.8)
```

File: README.md [hunk 3]
```diff
@@ -247,35 +295,80 @@ std::optional<lex_item> lexer(std::string_view v) noexcept {
...
+
+project(MyProject VERSION 1.0)
+set(CMAKE_CXX_STANDARD 20)
+set(CMAKE_CXX_STANDARD_REQUIRED True)
```
Comment (friendlyanon) [inline README.md:356]: CMake 3.8 does not know about C++20, you must not use this value. Please read the [docs](https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html):

> New in version 3.12

This is not the way to set the standard requirement of a target, you must use [compile features](https://cmake.org/cmake/help/latest/manual/cmake-compile-features.7.html#requiring-language-standards) for that.

File: README.md [hunk 4]
```diff
@@ -247,35 +295,80 @@ std::optional<lex_item> lexer(std::string_view v) noexcept {
...
+
+FetchContent_Declare(
+        ctre
+        GIT_REPOSITORY https://github.com/hanickadot/compile-time-regular-expressions.git
+        GIT_TAG 95c63867bf0f6497825ef6cf44a7d0791bd25883 # v3.4.1
+)
```
Comment (friendlyanon) [inline README.md:362]: The include is missing and please read the [docs](https://cmake.org/cmake/help/latest/module/FetchContent.html):

> New in version 3.11

File: README.md [hunk 5]
```diff
@@ -247,35 +295,80 @@ std::optional<lex_item> lexer(std::string_view v) noexcept {
...
+
+FetchContent_MakeAvailable(ctre)
```
Comment (friendlyanon) [inline README.md:364]: Please read the [docs](https://cmake.org/cmake/help/latest/module/FetchContent.html#command:fetchcontent_makeavailable):

> New in version 3.14

If https://github.com/hanickadot/compile-time-regular-expressions/pull/230 is merged then this becomes entirely wrong, as using ctre would simply be:

```cmake
find_package(ctre REQUIRED)
target_link_libraries(project_target PRIVATE ctre::ctre)
```

File: README.md [hunk 6]
```diff
@@ -247,35 +295,80 @@ std::optional<lex_item> lexer(std::string_view v) noexcept {
...
+include_directories("${ctre_SOURCE_DIR}/single-header")
```
Comment (friendlyanon) [inline README.md:365]: Don't use directory scope commands. Prefer [target_include_directories](https://cmake.org/cmake/help/latest/command/target_include_directories.html).

File: README.md [hunk 7]
```diff
@@ -247,35 +295,80 @@ std::optional<lex_item> lexer(std::string_view v) noexcept {
...
+
+# Add an executable with the above sources
+add_executable(${PROJECT_NAME} main.cpp)
+target_link_libraries(${PROJECT_NAME})
```
Comment (friendlyanon) [inline README.md:369]:
```suggestion
add_executable(example main.cpp)
target_link_libraries(example PRIVATE ctre::ctre)
```

File: README.md [hunk 8]
```diff
@@ -247,35 +295,80 @@ std::optional<lex_item> lexer(std::string_view v) noexcept {
...
+FetchContent_Declare(
+        ctre
+        GIT_REPOSITORY https://github.com/hanickadot/compile-time-regular-expressions.git
+        GIT_TAG 95c63867bf0f6497825ef6cf44a7d0791bd25883 # v3.4.1
+)
+
+FetchContent_MakeAvailable(ctre)
+include_directories("${ctre_SOURCE_DIR}/single-header")
+
+# Add an executable with the above sources
```
Comment (friendlyanon) [inline README.md:367]: > with the above sources

What sources?

**Review by friendlyanon — CHANGES_REQUESTED** [review]
The added CMake example has issues on almost every single line.

File: README.md (hunk 1 above)
Comment (angeletakis) [reply README.md:27]: I am using the github.com website that's why I created the ToC

File: README.md (hunk 4 above)
Comment (angeletakis) [reply README.md:362]: Oh I totally forgot that
include(FetchContent)

File: README.md (hunk 3 above)
Comment (angeletakis) [reply README.md:356]: Oh,

> For compilers that have no notion of a standard level, such as Microsoft Visual C++ before 2015 Update 3, this has no effect.

File: README.md (hunk 8 above)
Comment (angeletakis) [reply README.md:367]: Ctre

File: README.md (hunk 7 above)
Comment (angeletakis) [reply README.md:369]: I feel that the CMake example would be more of a template if we did
add_executable(${PROJECT_NAME} main.cpp)
target_link_libraries(${PROJECT_NAME} PRIVATE ctre::ctre)

File: README.md [hunk 9]
```diff
@@ -247,35 +295,80 @@ std::optional<lex_item> lexer(std::string_view v) noexcept {
...
+
+//using namespace ctre;
+```
+This header can be regenerated with `make single-header`.
+### CMake
+If you are using cmake, you can add this directory as subdirectory and link to target `ctre`.
+
+```CMake
+cmake_minimum_required(VERSION 3.8.0)
+
```
Comment (angeletakis) [inline README.md:353]:
```suggestion
include(FetchContent)
```

File: README.md (hunk 5 above)
Comment (angeletakis) [reply README.md:364]: I think that if the package was not found we could do

(pseudocode)
```
find_package(ctre)
if(package not found){
    FetchContent(ctre)
    include_target_directories(ctre dir)
}
```

Comment (friendlyanon) [reply README.md:365]: That's not the responsibility of this project. There are many ways to "polyfill" a `find_package` call. It should instead just use idiomatic CMake in examples.

File: README.md (hunk 7 above)
Comment (friendlyanon) [reply README.md:370]: Using `${PROJECT_NAME}` is pointless either way.

File: README.md (hunk 8 above)
Comment (friendlyanon) [reply README.md:368]: ctre is a header only library, it has no source files and to the headers to the include path you use `target_link_libraries(project_target PRIVATE ctre::ctre)` (once the appropriate PR is merged).

File: README.md (hunk 5 above)
Comment (angeletakis) [reply README.md:365]: I guess
```suggestion
find_package(ctre REQUIRED) 
target_link_libraries(project_target PRIVATE ctre::ctre)
```

File: README.md (hunk 4 above)
Comment (angeletakis) [reply README.md:363]:
```suggestion
```

File: README.md (hunk 7 above)
Comment (angeletakis) [reply README.md:370]: `${PROJECT_NAME}` makes it so there are fewer things to rewrite.

Comment (friendlyanon) [reply README.md:370]: Find and replace exists. How often does the project name change that this abstraction needs to exist? This also just obfuscates the target's name. There are only downsides to using this.

File: README.md (hunk 1 above)
Comment (friendlyanon) [reply README.md:27]: Yes, GitHub UI provides a ToC already:

image

Comment (angeletakis) [reply README.md:27]: Oh I didn't know that

File: README.md (hunk 7 above)
Comment (angeletakis) [reply README.md:370]: The project name changes every time someone copies the CMake example template and puts it in their own project. I would say that anyone wanting to not use the variable `${PROJECT_NAME}` is already further along in their project where they would only need to <kbd>Ctrl</kbd>+<kbd>C</kbd> the middle portion of the template.

Comment (friendlyanon) [reply README.md:370]: If this is just for a copy-paste example, then using an explicit name makes no semantic difference, but syntactically aligns better with what one should write to begin with. Anyone sufficiently knowledgeable regarding CMake will not bother with this example, as the only interesting thing is the exported target(s) in the install interface.

**Pull request closed without merging**